### PR TITLE
feat(response): allow to use a route identifier to redirect

### PR DESCRIPTION
**Pull request by RomainLanz**
Hey! 👋 

This PR allows us to use a route identifier to redirect using the `response.redirect()` helper.
The default behaviour is to try to resolve the given identifier as a route if no routes are found it fallback to the default behaviour and use the given string as an URL.

```ts
// Using "post.show" route name an adding params
response.redirect('post.show', { params: { id: 1 }})

// Using controller's method
response.redirect('PostsController.index')

// Specifying domain
response.redirect('post.create', {}, 302, 'domain.example.com')
```

Note that I added a test that is skipped because it tries to resolve the given domain, which makes the test fail. How can we avoid that?

### CODE REVIEW

File: src/Response/index.ts [hunk 1]
```diff
@@ -820,11 +822,33 @@ export class Response extends Macroable implements ResponseContract {
...
+    reflectQueryParams?: MakeUrlOptions | boolean,
+    statusCode: number = 302,
+    domain?: string,
+  ): void {
+    const route = this.router.lookup(routerIdentifierOrUrl, domain)
+
+    if (route) {
+      const url = this.router.makeUrl(routerIdentifierOrUrl, reflectQueryParams as MakeUrlOptions, domain) as string
+
+      return this.redirectToPath(url, false, statusCode)
```
Comment (thetutlage) [inline src/Response/index.ts:841]: Why we are not allowing passing query params to be redirect here?

**Comment by RomainLanz** [comment]
**Updates**

We new provide a `Redirect` class that handles everything about redirection.

Signature:
```ts
export interface RedirectContract {
  status (statusCode: number): this
  withQs (): this
  withQs (values: { [key: string]: any }): this
  withQs (name: string, value: any): this

  back (): void
  toRoute (routeIdentifier: string, urlOptions?: MakeUrlOptions, domain?: string): void
  toPath (url: string): void
}
```

You can now pass 0 argument to the `response.redirect()` method to get an instance of the class.

```ts
response.redirect().toRoute('PostsController.index')
response.redirect().back()
response.redirect().withQs('username', 'romain').toPath('foo')
// ...
```

File: adonis-typings/redirect.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+/*
+ * @adonisjs/http-server
+ *
+ * (c) Harminder Virk <virk@adonisjs.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare module '@ioc:Adonis/Core/Redirect' {
```
Comment (thetutlage) [inline adonis-typings/redirect.ts:10]: Why put it inside it's own module? I believe declaring the interface within the response module is fine?

File: src/Redirect/index.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+  constructor (
+    private request: IncomingMessage,
+    private response: ResponseContract,
+    private router: RouterContract
+  ) {
+  }
+
+  private computeQueryString (): string {
+    return Array.from(this.queryString.entries())
```
Comment (thetutlage) [inline src/Redirect/index.ts:33]: We should use [qs.stringify](https://www.npmjs.com/package/qs), it generates a spec complaint query string

**Comment by thetutlage** [comment]
Beautiful ❤️

**Pull request merged**